learning/tour-of-beam/frontend/lib/pages/tour/state.dart

```diff
@@ -55,10 +61,66 @@ class TourNotifier extends ChangeNotifier with PageStateMixin<void> {
     contentTreeController.addListener(_onUnitChanged);
     _unitContentCache.addListener(_onUnitChanged);
     _appNotifier.addListener(_onAppNotifierChanged);
-    _authNotifier.addListener(_onUnitProgressChanged);
+    _authNotifier.addListener(_onAuthChanged);
+    _saveCodeDebounced = _saveUserCode.debounced(
+      _saveUserCodeDebounceDuration,
+    );
+    _setSaveCodeListener();
     _onUnitChanged();
   }
 
+  void _setSaveCodeListener() {
+    // Creates snippetEditingController
+    playgroundController.setSdk(_appNotifier.sdk!);
+    // Notifies when activeFileController is created
+    playgroundController.snippetEditingController?.addListener(
+      _snippedEditingControllerListener,
+    );
+  }
+
+  void _snippedEditingControllerListener() {
+    // Notifies when code is changed
+    playgroundController
+        .snippetEditingController?.activeFileController?.codeController
+        .addListener(_saveCodeListener);
+  }
+
+  void _saveCodeListener() {
+    final currentUnit = currentUnitController;
+    final isCodeChanged = playgroundController
+            .snippetEditingController?.activeFileController?.isChanged ??
+        false;
+    final code = playgroundController.snippetEditingController
+        ?.activeFileController?.codeController.fullText;
+    final doSave = _authNotifier.isAuthenticated &&
+        isCodeChanged &&
+        currentUnit != null &&
+        code != null;
+
+    if (doSave) {
+      _saveCodeDebounced?.call([
+        currentUnit.sdkId,
+        currentUnit.unitId,
+        code,
+      ]);
+    }
+  }
+
+  Future<void> _saveUserCode(String sdkId, String unitId, String code) async {
+    try {
+      final client = GetIt.instance.get<TobClient>();
+      await client.postUserCode(sdkId: sdkId, unitId: unitId, code: code);
+      if (!hasSavedSnippet) {
+        await _unitProgressCache.updateUnitProgress();
+        _snippetType = SnippetType.saved;
+        notifyListeners();
+      }
+    } on Exception catch (e) {
+      // TODO(nausharipov): how to handle?
```

Comment:
Add 'Saving...' indicator. Light it when the snippet is changed. Turn it off when the user snippet matches the content.